repose-aggregator/services/atom-feed/impl/src/main/scala/org/openrepose/nodeservice/atomfeed/impl/AtomFeedServiceImpl.scala

```diff
@@ -0,0 +1,228 @@
+          feedListeners = feedListeners + (feedId -> (pastListeners + (listenerId -> listener)))
+          if (pastListeners.isEmpty) {
+            val schedule = actorSystem.scheduler.schedule(
+              Duration.Zero,
+              feedConfig.getPollingFrequency.seconds,
+              actorSystem.actorOf(FeedReader.props(
+                feedConfig.getUri,
+                (arf: ActorRefFactory) =>
+                  arf.actorOf(Authenticator.props(feedConfig.getAuthentication)),
+                (arf: ActorRefFactory) =>
+                  arf.actorOf(Notifier.props(feedListeners.getOrElse(feedId, Map.empty).values.toSet)),
+                feedConfig.getEntryOrder
+              )),
+              ReadFeed
+            )
+            feedSchedules = feedSchedules + (feedId -> schedule)
+          }
+
+          listenerId
+        case None =>
+          throw new IllegalArgumentException("feedId parameter does not match a configured feed id")
+      }
+    }
+  }
+
+  override def unregisterListener(listenerId: String): Unit = {
+    ifRunning {
+      feedListeners find {
+        case (_, listeners) => listeners.contains(listenerId)
+      } match {
+        case Some((feedId, listeners)) =>
+          logger.debug("Un-registering the listener: " + listenerId)
+
+          val newListeners = listeners filterNot { case (id, listener) => listenerId.equals(id) }
+          feedListeners = feedListeners + (feedId -> newListeners)
+          if (newListeners.isEmpty) {
+            feedSchedules(feedId).cancel()
+          }
+        case None =>
+          logger.debug("Listener not registered: " + listenerId)
+      }
+    }
+  }
+
+  override def isRunning: Boolean = running
+
+  private def startService(): Unit = {
+    synchronized {
+      if (isInitialized && isServiceEnabled && !running) {
+        logger.info("Starting the service")
+        actorSystem = ActorSystem.create("AtomFeedServiceSystem")
+        actorSystem.registerOnTermination(running = false)
+        running = true
+      } else if (running) {
+        logger.debug("Service not starting -- service is already running")
+      } else if (!isServiceEnabled) {
+        logger.debug("Service not starting -- disabled by configuration")
+      } else {
+        logger.debug("Service not starting -- waiting on configuration files to load")
+      }
+    }
+  }
+
+  private def stopService(): Unit = {
+    synchronized {
+      if (running) {
+        logger.info("Stopping the service")
+        actorSystem.shutdown()
+        actorSystem.awaitTermination(5 seconds)
+      } else {
+        logger.debug("Service not stopping -- service is already stopped")
+      }
+    }
+  }
+
+  private def ifRunning[T](f: => T): T = {
+    if (running) f
+    else throw new IllegalStateException("Service is not running")
+  }
+
+  private def getListenersForFeed(feedId: String): Map[String, AtomFeedListener] = {
```

Comment:
So this is kind of a hacky thing I did. Part of letting the service itself maintain the listener mapping is providing actors a way to access that mapping. We can't pass the mapping in when we construct an actor because it changes over time. To get around that, I pass a function with a reference to the mapping that the actors can use. So this does actually get called a bunch.

Originally, I just had the service sending messages to actors whenever a change to listeners came in. That approach led to a bunch of cookie-cutter code to forward a call to registerListener into the FeedReader actor into the Notifier actor (where the mapping was finally stored/updated). Admittedly, that may be the right way to do it, but I figured I could save some overhead by passing around a function instead.